domains/client/domain-operator/src/operator.rs

```diff
@@ -178,19 +191,167 @@ where
+                    if let Some(ref snap_sync_orchestrator) = snap_sync_orchestrator {
+                        let mut target_block_receiver =
+                            snap_sync_orchestrator.consensus_snap_sync_target_block_receiver();
+                        let target_block_result = target_block_receiver.recv().await;
+
+                        let target_block_number =
+                            if let Ok(target_block) = target_block_result.clone() {
+                                target_block
+                            } else {
+                                error!(
+                                    ?target_block_result,
+                                    "Snap sync failed: can't obtain target block."
+                                );
+
+                                return Err(sp_consensus::Error::Other(
+                                    "Snap sync failed: can't obtain target block.".into(),
+                                ));
+                            };
+
+                        // Wait for Subspace block importing notifications
+                        let mut block_importing_notification_stream =
+                            Box::pin(params.operator_streams.block_importing_notification_stream);
+
+                        while let Some((block_number, mut acknowledgement_sender)) =
+                            block_importing_notification_stream.next().await
+                        {
+                            trace!(%block_number, "Acknowledged block import from consensus chain.");
+                            if acknowledgement_sender.send(()).await.is_err() {
+                                return Err(sp_consensus::Error::Other(
+                                    format!("Can't acknowledge block import #{}", block_number)
+                                        .into(),
+                                ));
+                            }
+
+                            if block_number >= target_block_number.into() {
+                                break;
+                            }
+                        }
+
+                        // Drain Substrate block imported notifications
+                        let mut imported_block_notification_stream =
+                            Box::pin(params.operator_streams.imported_block_notification_stream);
+
+                        while let Some(import_notification) =
+                            imported_block_notification_stream.next().await
+                        {
+                            let block_number = *import_notification.header.number();
+                            trace!(%block_number, "Block imported from consensus chain.");
+
+                            if block_number >= target_block_number.into() {
+                                break;
+                            }
+                        }
+
+                        // Restore parameters
+                        unsafe {
```

Comment:
I think there’s a way to avoid this unsafe block by using a mutable reference as the boxed stream. I’ll try some things and follow up with a specific suggestion.

If we do need this unsafe block, it would be good to have a safety comment that explains how we satisfy the specific safety requirements of the unsafe methods used.

Reply:
Hmm, it's a bit more complicated than I thought. But the underlying type `TracingUnboundedReceiver` is `Unpin`, so we might be able to fix it by adding an `Unpin` bound to the generics on this method and its callers.

I'll work on it more tomorrow, and submit a PR if I can.

Reply:
Yep, we just need to promise the streams are pinned. This is already true for the underlying types, but in some cases we need to re-pin the streams.

See the details and code in PR #3179.

Reply:
The refactored code worked well. It's much cleaner now Thanks!